// @progress/kendo-angular-progressbar/esm2020/progressbar.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChunkProgressBarComponent } from './chunk/chunk-progressbar.component';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ProgressBarComponent } from './progressbar.component';
import { ResizeSensorModule } from '@progress/kendo-angular-common';
import { CircularProgressBarComponent } from './circularprogressbar/circular-progressbar.component';
import { CircularProgressbarCenterTemplateDirective } from './circularprogressbar/center-template.directive';
import { LocalizedProgressBarMessagesDirective } from './common/localization/localized-messages.directive';
import { ProgressBarCustomMessagesComponent } from './common/localization/custom-messages.component';
import * as i0 from "@angular/core";
const COMPONENT_DIRECTIVES = [
    ProgressBarComponent,
    ChunkProgressBarComponent,
    CircularProgressBarComponent,
    CircularProgressbarCenterTemplateDirective,
    LocalizedProgressBarMessagesDirective,
    ProgressBarCustomMessagesComponent
];
const MODULES = [CommonModule, ResizeSensorModule];
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmodules'])
 * definition for the ProgressBar components.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the ProgressBar module
 * import { ProgressBarModule } from '@progress/kendo-angular-progressbar';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, ProgressBarModule], // import ProgressBar module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
export class ProgressBarModule {
}
ProgressBarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: ProgressBarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ProgressBarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: ProgressBarModule, declarations: [ProgressBarComponent,
        ChunkProgressBarComponent,
        CircularProgressBarComponent,
        CircularProgressbarCenterTemplateDirective,
        LocalizedProgressBarMessagesDirective,
        ProgressBarCustomMessagesComponent], imports: [CommonModule, ResizeSensorModule], exports: [ProgressBarComponent,
        ChunkProgressBarComponent,
        CircularProgressBarComponent,
        CircularProgressbarCenterTemplateDirective,
        LocalizedProgressBarMessagesDirective,
        ProgressBarCustomMessagesComponent] });
ProgressBarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: ProgressBarModule, imports: [MODULES] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: ProgressBarModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: COMPONENT_DIRECTIVES,
                    exports: COMPONENT_DIRECTIVES,
                    imports: MODULES
                }]
        }] });
