// @progress/kendo-angular-schematics/src/third_party/index.js
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addDeclarationToNgModule = exports.buildSelector = exports.strings = exports.Style = exports.validateHtmlSelector = exports.parseName = exports.findModuleFromOptions = exports.getWorkspace = exports.buildDefaultPath = void 0;
const tslib_1 = require("tslib");
const package_json_1 = require("@schematics/angular/package.json");
const semver_1 = tslib_1.__importDefault(require("semver"));
var workspace_1 = require("@schematics/angular/utility/workspace");
Object.defineProperty(exports, "buildDefaultPath", { enumerable: true, get: function () { return workspace_1.buildDefaultPath; } });
Object.defineProperty(exports, "getWorkspace", { enumerable: true, get: function () { return workspace_1.getWorkspace; } });
var find_module_1 = require("@schematics/angular/utility/find-module");
Object.defineProperty(exports, "findModuleFromOptions", { enumerable: true, get: function () { return find_module_1.findModuleFromOptions; } });
var parse_name_1 = require("@schematics/angular/utility/parse-name");
Object.defineProperty(exports, "parseName", { enumerable: true, get: function () { return parse_name_1.parseName; } });
var validation_1 = require("@schematics/angular/utility/validation");
Object.defineProperty(exports, "validateHtmlSelector", { enumerable: true, get: function () { return validation_1.validateHtmlSelector; } });
var schema_1 = require("@schematics/angular/component/schema");
Object.defineProperty(exports, "Style", { enumerable: true, get: function () { return schema_1.Style; } });
;
exports.strings = semver_1.default.satisfies(package_json_1.version, "<14.0.0")
    ? require("@angular-devkit/core").strings
    : require("@angular-devkit/schematics").strings;
function buildSelector(options, projectPrefix) {
    let selector = exports.strings.dasherize(options.name);
    if (options.prefix) {
        selector = `${options.prefix}-${selector}`;
    }
    else if (options.prefix === undefined && projectPrefix) {
        selector = `${projectPrefix}-${selector}`;
    }
    return selector;
}
exports.buildSelector = buildSelector;
exports.addDeclarationToNgModule = semver_1.default.satisfies(package_json_1.version, "<15.0.0")
    ? require("./index.v13").addDeclarationToNgModule
    : require("@schematics/angular/utility/add-declaration-to-ng-module").addDeclarationToNgModule;
