// @progress/kendo-angular-layout/esm2020/avatar.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { IconsModule } from '@progress/kendo-angular-icons';
import { AvatarComponent } from './avatar/avatar.component';
import { LocalizedAvatarMessagesDirective } from './avatar/l10n/localized-messages.directive';
import { AvatarCustomMessagesComponent } from './avatar/l10n/custom-messages.component';
import * as i0 from "@angular/core";
const exportedModules = [
    AvatarComponent,
    AvatarCustomMessagesComponent
];
const declarations = [
    LocalizedAvatarMessagesDirective,
    ...exportedModules
];
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the Avatar component.
 */
export class AvatarModule {
}
AvatarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: AvatarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AvatarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: AvatarModule, declarations: [LocalizedAvatarMessagesDirective, AvatarComponent,
        AvatarCustomMessagesComponent], imports: [CommonModule, IconsModule], exports: [AvatarComponent,
        AvatarCustomMessagesComponent] });
AvatarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: AvatarModule, imports: [[CommonModule, IconsModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: AvatarModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [declarations],
                    exports: [exportedModules],
                    imports: [CommonModule, IconsModule]
                }]
        }] });
