// @progress/kendo-angular-layout/esm2020/splitter.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DraggableModule } from '@progress/kendo-angular-common';
import { SplitterComponent } from './splitter/splitter.component';
import { SplitterBarComponent } from './splitter/splitter-bar.component';
import { SplitterPaneComponent } from './splitter/splitter-pane.component';
import { IconsModule } from '@progress/kendo-angular-icons';
import * as i0 from "@angular/core";
const exportedModules = [
    SplitterComponent,
    SplitterPaneComponent
];
const declarations = [
    SplitterBarComponent,
    ...exportedModules
];
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the Splitter component.
 *
 * The module registers:
 * - `SplitterComponent`&mdash;The `Splitter` component class.
 * - `SplitterPaneComponent`&mdash;The `SplitterPane` component class.
 */
export class SplitterModule {
}
SplitterModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: SplitterModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SplitterModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: SplitterModule, declarations: [SplitterBarComponent, SplitterComponent,
        SplitterPaneComponent], imports: [CommonModule,
        DraggableModule,
        IconsModule], exports: [SplitterComponent,
        SplitterPaneComponent] });
SplitterModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: SplitterModule, imports: [[
            CommonModule,
            DraggableModule,
            IconsModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: SplitterModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [declarations],
                    exports: [exportedModules],
                    imports: [
                        CommonModule,
                        DraggableModule,
                        IconsModule
                    ]
                }]
        }] });
