// @progress/kendo-angular-layout/esm2020/stepper.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ProgressBarModule } from '@progress/kendo-angular-progressbar';
import { StepperComponent } from './stepper/stepper.component';
import { StepperStepComponent } from './stepper/step.component';
import { StepperListComponent } from './stepper/list.component';
import { StepperIndicatorTemplateDirective } from './stepper/template-directives/indicator-template.directive';
import { StepperLabelTemplateDirective } from './stepper/template-directives/label-template.directive';
import { StepperStepTemplateDirective } from './stepper/template-directives/step-template.directive';
import { LocalizedStepperMessagesDirective } from './stepper/localization/localized-messages.directive';
import { StepperCustomMessagesComponent } from './stepper/localization/custom-messages.component';
import { IconsModule } from '@progress/kendo-angular-icons';
import * as i0 from "@angular/core";
const templateDirectives = [
    StepperStepTemplateDirective,
    StepperLabelTemplateDirective,
    StepperIndicatorTemplateDirective
];
const exportedModules = [
    StepperComponent,
    StepperCustomMessagesComponent,
    ...templateDirectives
];
const declarations = [
    StepperStepComponent,
    StepperListComponent,
    LocalizedStepperMessagesDirective,
    ...exportedModules
];
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the Stepper component.
 */
export class StepperModule {
}
StepperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: StepperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
StepperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: StepperModule, declarations: [StepperStepComponent,
        StepperListComponent,
        LocalizedStepperMessagesDirective, StepperComponent,
        StepperCustomMessagesComponent, StepperStepTemplateDirective,
        StepperLabelTemplateDirective,
        StepperIndicatorTemplateDirective], imports: [CommonModule, ProgressBarModule, IconsModule], exports: [StepperComponent,
        StepperCustomMessagesComponent, StepperStepTemplateDirective,
        StepperLabelTemplateDirective,
        StepperIndicatorTemplateDirective] });
StepperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: StepperModule, imports: [[CommonModule, ProgressBarModule, IconsModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: StepperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [declarations],
                    exports: [exportedModules],
                    imports: [CommonModule, ProgressBarModule, IconsModule]
                }]
        }] });
