// @progress/kendo-angular-layout/esm2020/tabstrip.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { TabComponent } from './tabstrip/rendering/tab.component';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TabStripComponent } from './tabstrip/tabstrip.component';
import { TabStripTabComponent } from './tabstrip/models/tabstrip-tab.component';
import { TabContentDirective } from './tabstrip/directives/tab-content.directive';
import { TabTitleDirective } from './tabstrip/directives/tab-title.directive';
import { TabStripCustomMessagesComponent } from './tabstrip/localization/custom-messages.component';
import { LocalizedTabStripMessagesDirective } from './tabstrip/localization/localized-messages.directive';
import { ResizeSensorModule, WatermarkModule } from '@progress/kendo-angular-common';
import { TabStripScrollableButtonComponent } from './tabstrip/scrollable-button.component';
import { IconsModule } from '@progress/kendo-angular-icons';
import { ButtonModule } from '@progress/kendo-angular-buttons';
import { TabTemplateDirective } from './tabstrip/directives/tab.directive';
import * as i0 from "@angular/core";
const exportedModules = [
    TabStripComponent,
    TabStripTabComponent,
    TabContentDirective,
    TabTitleDirective,
    TabComponent,
    TabStripCustomMessagesComponent,
    LocalizedTabStripMessagesDirective,
    TabTemplateDirective
];
const declarations = [
    ...exportedModules,
    TabStripScrollableButtonComponent
];
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the TabStrip component.
 *
 * The module registers:
 * - `TabStripComponent`&mdash;The `TabStrip` component class.
 * - `TabStripTabComponent`&mdash;The `TabStripTab` component class.
 * - `TabContentDirective`&mdash;The tab content directive used on the `<ng-template>` tag.
 * - `TabTitleDirective`&mdash;The tab title directive used on the `<ng-template>` tag.
 */
export class TabStripModule {
}
TabStripModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: TabStripModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TabStripModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: TabStripModule, declarations: [TabStripComponent,
        TabStripTabComponent,
        TabContentDirective,
        TabTitleDirective,
        TabComponent,
        TabStripCustomMessagesComponent,
        LocalizedTabStripMessagesDirective,
        TabTemplateDirective, TabStripScrollableButtonComponent], imports: [CommonModule, ResizeSensorModule, IconsModule, ButtonModule, WatermarkModule], exports: [TabStripComponent,
        TabStripTabComponent,
        TabContentDirective,
        TabTitleDirective,
        TabComponent,
        TabStripCustomMessagesComponent,
        LocalizedTabStripMessagesDirective,
        TabTemplateDirective] });
TabStripModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: TabStripModule, imports: [[CommonModule, ResizeSensorModule, IconsModule, ButtonModule, WatermarkModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: TabStripModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [declarations],
                    exports: [exportedModules],
                    imports: [CommonModule, ResizeSensorModule, IconsModule, ButtonModule, WatermarkModule]
                }]
        }] });
