// @progress/kendo-angular-layout/esm2020/tilelayout.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { WatermarkModule } from '@progress/kendo-angular-common';
import { TileLayoutComponent } from './tilelayout/tilelayout.component';
import { TileLayoutItemComponent } from './tilelayout/tilelayout-item.component';
import { TileLayoutItemHeaderComponent } from './tilelayout/tilelayout-item-header.component';
import { TileLayoutItemBodyComponent } from './tilelayout/tilelayout-item-body.component';
import { TileLayoutResizeHandleDirective } from './tilelayout/tilelayout-resize-handle.directive';
import * as i0 from "@angular/core";
const exportedModules = [
    TileLayoutComponent,
    TileLayoutItemComponent,
    TileLayoutItemHeaderComponent,
    TileLayoutItemBodyComponent,
    TileLayoutResizeHandleDirective
];
const declarations = [
    ...exportedModules
];
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the TileLayout component.
 *
 * The module registers:
 * - `TileLayoutComponent`&mdash;The `TileLayoutComponent` component class.
 * - `TileLayoutItemComponent`&mdash;The `TileLayoutItemComponent` component class.
 * - `TileLayoutItemHeaderComponent`&mdash;The `TileLayoutItemHeaderComponent` component class.
 * - `TileLayoutItemBodyComponent`&mdash;The `TileLayoutItemBodyComponent` component class.
 */
export class TileLayoutModule {
}
TileLayoutModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: TileLayoutModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TileLayoutModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: TileLayoutModule, declarations: [TileLayoutComponent,
        TileLayoutItemComponent,
        TileLayoutItemHeaderComponent,
        TileLayoutItemBodyComponent,
        TileLayoutResizeHandleDirective], imports: [CommonModule, WatermarkModule], exports: [TileLayoutComponent,
        TileLayoutItemComponent,
        TileLayoutItemHeaderComponent,
        TileLayoutItemBodyComponent,
        TileLayoutResizeHandleDirective] });
TileLayoutModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: TileLayoutModule, imports: [[CommonModule, WatermarkModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: TileLayoutModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [declarations],
                    exports: [exportedModules],
                    imports: [CommonModule, WatermarkModule]
                }]
        }] });
