// @progress/kendo-angular-common/esm2020/adornments/prefix.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, Input, Optional, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Specifies the adornments in the prefix container of the [Inputs]({% slug adornments_textbox %}#toc-prefixadornments) and [DropDowns]({% slug adornments_multiselect %}#toc-prefixadornments).
 * ```ts-no-run
 * _@Component({
 * selector: 'my-app',
 * template: `
 *  <kendo-textbox>
 *      <ng-template kendoPrefixTemplate>
 *          <button kendoButton look="clear" icon="image"></button>
 *      </ng-template>
 *  </kendo-textbox>
 *
 *  <kendo-multiselect [data]="data" [(ngModel)]="value">
 *      <ng-template kendoPrefixTemplate>
 *          <button kendoButton look="clear" icon="image"></button>
 *      </ng-template>
 *  </kendo-multiselect>
 * `
 * })
 * class AppComponent {}
 * ```
 */
export class PrefixTemplateDirective {
    constructor(templateRef) {
        this.templateRef = templateRef;
        this._showSeparator = false;
    }
    /**
     * Sets the `showSeparator` attribute of the `prefixTemplate`.
     *
     * @default false
     */
    set showSeparator(value) {
        this._showSeparator = value;
    }
    get showSeparator() {
        return this._showSeparator;
    }
}
PrefixTemplateDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: PrefixTemplateDirective, deps: [{ token: i0.TemplateRef, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
PrefixTemplateDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.12", type: PrefixTemplateDirective, selector: "[kendoPrefixTemplate]", inputs: { showSeparator: "showSeparator" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: PrefixTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoPrefixTemplate]'
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef, decorators: [{
                    type: Optional
                }] }]; }, propDecorators: { showSeparator: [{
                type: Input
            }] } });
