// @progress/kendo-angular-common/esm2020/watermark/watermark.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2024 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { WatermarkOverlayComponent } from './watermark.component';
import * as i0 from "@angular/core";
/**
 * @hidden
 */
export class WatermarkModule {
}
WatermarkModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: WatermarkModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
WatermarkModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: WatermarkModule, declarations: [WatermarkOverlayComponent], imports: [CommonModule], exports: [WatermarkOverlayComponent] });
WatermarkModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: WatermarkModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: WatermarkModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [WatermarkOverlayComponent],
                    exports: [WatermarkOverlayComponent],
                    imports: [CommonModule]
                }]
        }] });
